test/util/clienthelper/hookclient.go

```diff
@@ -18,59 +18,110 @@ type hookFunc func(obj client.Object) error
 type HookingClient struct {
 	f client.WithWatch
 
-	getHook    []getFunc
-	deleteHook []hookFunc
-	createHook []hookFunc
-	updateHook []hookFunc
-	patchHook  []hookFunc
+	preGetHook    []getFunc
+	preDeleteHook []hookFunc
+	preCreateHook []hookFunc
+	preUpdateHook []hookFunc
+	prePatchHook  []hookFunc
+
+	postGetHook    []getFunc
+	postDeleteHook []hookFunc
+	postCreateHook []hookFunc
+	postUpdateHook []hookFunc
+	postPatchHook  []hookFunc
 }
 
 var _ client.Client = &HookingClient{}
 
-// NewHookingClient creates a
+// NewHookingClient creates a client which allows hooks to be added before and
+// after the client event. Errors returned in the hooks are returned to the
+// caller directly, to simulate issues such as disconnections or errors.
+// Prehooks that cause errors will cause the underlying wrapped client to not be
+// called.
 func NewHookingClient(c client.WithWatch) *HookingClient {
 	return &HookingClient{
-		f:          c,
-		getHook:    []getFunc{},
-		deleteHook: []hookFunc{},
-		createHook: []hookFunc{},
-		updateHook: []hookFunc{},
-		patchHook:  []hookFunc{},
+		f:             c,
+		preGetHook:    []getFunc{},
+		preDeleteHook: []hookFunc{},
+		preCreateHook: []hookFunc{},
+		preUpdateHook: []hookFunc{},
+		prePatchHook:  []hookFunc{},
+
+		postGetHook:    []getFunc{},
+		postDeleteHook: []hookFunc{},
+		postCreateHook: []hookFunc{},
+		postUpdateHook: []hookFunc{},
+		postPatchHook:  []hookFunc{},
 	}
 }
 
 func (c *HookingClient) WithGetHook(f getFunc) *HookingClient {
-	c.getHook = append(c.getHook, f)
+	c.postGetHook = append(c.postGetHook, f)
 	return c
 }
 
 func (c *HookingClient) WithDeleteHook(f hookFunc) *HookingClient {
```

Comment:
If we have `WithPreDeleteHook` I would be in favour of `WithPostDeleteHook` for symmetry and discoverability. Same with the others here.

What do you think?

Reply:
I agree, doesn't block PR though

Reply:
PR for it here: https://github.com/Azure/ARO-RP/pull/3549